### DISCUSSION TITLE
Enhancement: Switching between Chat View and MindMap View

### DISCUSSION THREAD

**Discussion by ryohei-kamiya** [danny-avila/LibreChat#1655] [Feature Requests & Suggestions]
### What features would you like to see added?

The concept of Enhancement in this project is great and I really like it.
I hope that my proposal in below is in line with the concept and will contribute to further enhancement of the project.

When using AI chat tools like ChatGPT for brainstorming new content ideas (such as novel or comic storylines), I've encountered some inconveniences:

- Only a limited number of specific ideas can be generated at once.
- Increasing the number of ideas tends to simplify the content of those ideas.
- It's difficult to compare multiple ideas side by side on a single screen.

Personally, I think "MindMap" are well-suited to address these issues. Therefore, I suggest adding a MindMap view.

### More details

- Pressing the "Switch to MindMap View" button while the chat screen is open will switch the view from chat to a MindMap.
- When the MindMap view is open, pressing the "Switch to Chat View" button will change the MindMap back to the chat screen.
- Each node in the MindMap will contain two messages:
    - The root node will contain only one "user message."
    - Other nodes will each contain one "assistant message" and one "user message," in that order.
- Forms will be placed on each node in the MindMap to accept input from the user.
- Enlarged nodes will also contain a form, from which users can send messages.
- It is possible to receive multiple assistant messages in response to a single user message.
- User messages entered into forms on multiple nodes can be sent in parallel.
- The MindMap can be zoomed in and out as a whole, and nodes can be moved around for rearrangement.
- Clicking on a node with a mouse or tapping on it on a smartphone screen will enlarge the node for a closer view.
- Pressing the ESC key or the Close button will close the enlarged node view (the data entered in the form will be synchronized and remain in the form of the node before it was enlarged).

These features are already implemented in a branch of the forked repository below.

https://github.com/ryohei-kamiya/LibreChat/tree/feat-mindmap

For the implementation of MindMap, ReactFlow (https://reactflow.dev/) was used.

### Remaining Issues
- The feature that allows returning multiple responses to a single request currently only functions with OpenAI's API. This capability has not been extended to other APIs.
    - I've adapted existing APIs to deliver 'n' responses when the 'n' parameter is included in a request. If 'n' is not specified or set to 1, the system will return a single response.
        - Separating APIs based on differing response formats might be more logical, but such a division would significantly impact the frontend. To maintain efficiency and minimize extensive changes, I have maintained the current implementation strategy.
    - The challenge of implementing a streaming response for multiple messages with OpenAI's API led to a design where all messages are returned simultaneously after generation. A specification that allows for multiple reliable streams would be ideal if feasible.
- The button for toggling between views is located in the top left corner, which may not be the most optimal position. If a better location is identified, the button should be relocated.
- The state management for each MindMap node, particularly for submissions, has become complex.
- To keep changes to the existing code base to a minimum, I replicated existing frontend files under new names and applied modifications to these copies. This approach nearly doubled the number of frontend files. Redundant code needs to be reduced.
- I aim to give users the ability to adjust the number of responses per request in the MindMap view, which is currently fixed at three. Introducing a button on the MindMap node headers for adjusting the response count is a good idea, but the complexity it adds to state management has prevented its implementation.
- Keeping up with the rapid development and frequent updates of the original repository, integrating new changes is a huge challenge for me. The last update included was 5730028b8389100b98cf6e8823b2d66dc0edc9f0, and subsequent updates must be merged.
- Sending images from MindMap nodes sometimes causes responses to be truncated. The reason for this issue remains unclear.
- In the current setup, dispatching multiple requests concurrently can result in some responses not being displayed on the screen. Reloading the browser reveals the missing responses. Despite recognizing this issue, I haven't identified a straightforward fix, so the problem persists.
- The functionality has been tested using a browser(Chrome), but I haven't write any test code.

### Which components are impacted by your request?

UI

### Pictures

https://github.com/danny-avila/LibreChat/assets/2719533/aa5caddd-d7cb-4b27-b070-f56ab80d66dc

↓ Long version
https://drive.google.com/file/d/1fbcVebsAGbZpgwUy3Ledyfubx6GtoWa0/view?usp=sharing

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Wow, incredible work, thank you for sharing! This is one of the most comprehensive forks I've seen. Thank you for taking the time to dig into the code and making something truly unique. I will like to try it out and maybe make a video about it on YouTube, so others are aware and can provide feedback on what you've made.

There is a lot to digest here, and I hope to address all your points. For now, these points stand out to me:

> 1. The challenge of implementing a streaming response for multiple messages with OpenAI's API led to a design where all messages are returned simultaneously after generation. A specification that allows for multiple reliable streams would be ideal if feasible.
> 2. To keep changes to the existing code base to a minimum, I replicated existing frontend files under new names and applied modifications to these copies. This approach nearly doubled the number of frontend files. Redundant code needs to be reduced.
> 3. In the current setup, dispatching multiple requests concurrently can result in some responses not being displayed on the screen. Reloading the browser reveals the missing responses. Despite recognizing this issue, I haven't identified a straightforward fix, so the problem persists.

When I refactored a lot of the frontend code to the "new" (11/06/23) style of ChatGPT, I changed a lot of the conversation/submission logic for the very purpose of multiple streams/responses at once. 

I apologize that I haven't gotten around to multiple streams yet, as I think this could've simplified your approach and had some of your desired effects already. 

In other words, streaming multiple responses should be more than possible without much introduction of code or addition of files anywhere in the current codebase. This is made possible by the recoil atom families the app now uses which allows many of the states to be handled on an index basis. A way to handle multiple server-sent events in one view is yet to be implemented but is one example of something that would need to be introduced.

In conclusion, the mind map feature should be tackled when I've added multiple response streaming. The work you've done so far can mostly be used as reference when the update is in place, because I foresee under 20 files worth of changes for adding multiple response streaming on my end (and this is a conservative estimate, as could be lower). 

I can't merge your current work without a considerable amount of testing, and even then, the redundant code would bloat the frontend bundle size more than I can allow. Due to my limited time and current focus on the project, this is not currently feasible for me to address personally.

My last point is that the core use case for LibreChat is for the user accustomed to a simpler, familiar interface. As an AI enthusiast, I am open to features like MindMap, and integrating more knowledge-graph-like features is a strong personal desire, but it would need to be far more segmented than it currently is. In an upcoming update, I am introducing an optional "side panel", maybe enabling of the feature can live there.

I highly recommend you join the discord channel so we can communicate in real-time, this is highly encouraged for anyone making significant contributions: https://discord.librechat.ai

**Comment by ryohei-kamiya** [comment]
@danny-avila 
Thank you for your swift response! 
I'm also grateful for your offer to feature my work on your YouTube channel.
I would be delighted if you could showcase it, allowing many people to discover the potential of "MindMap View".

It's great news for me that the ability to handle multiple streams/responses is already in place.  I'm not well-versed in React and Recoil, which made it challenging for me to understand this multi-stream/response feature. I thought the 'index' within useChatHelpers might be the key, but lacking confidence, I decided to introduce an additional parameter 'nodeId' for my implementation. (So far, my coding experience has predominantly been with C, Python, and TypeScript(but Angular))

I understand that integrating the current code with the original code poses significant challenges, and I have no intention of complicating the project's roadmap.

Therefore, I agree with your decision to prioritize the original direction of the project. I would be very pleased if there's an opportunity to incorporate my suggestions into the "side panel" in the future.

Thank you for inviting me to the Discord channel!
I'm not very good at speaking English, so I expect our communication will mainly be text-based, but I'm looking forward to participating.
It's 2 AM here in Tokyo, and I'm feeling sleepy. I'll join the Discord during the day tomorrow. Thank you!

**Reply by rnehrboss** [reply]
Excited to see this.   Any idea when we'd see the fork merged?

**Reply by ryohei-kamiya** [reply]
@rnehrboss 
Thank you for your interest! The timing for the merge has not been decided yet, but I am hoping it is under consideration. Currently, the existing fork has too many additional files to merge, and the impact on maintenance after merging is significant.

I am continuing to explore ways to reduce the code.

**Reply by ryohei-kamiya** [reply]
I am continuously working on this. I'll share the current status. I've reworked the implementation to focus exclusively on the mind map display feature. In the previous implementation, there were over 200 files for new additions and edits, but now it's been reduced to about 15 files. But, as a trade-off, the chat input field is hidden when displaying the mind map.

I think the merge might be a bit delayed since the reviewers seem busy with other developments, but I'll continue to make progress on the development.

https://github.com/ryohei-kamiya/LibreChat/tree/feat-mindmap-1


https://github.com/danny-avila/LibreChat/assets/2719533/c19bb1fd-4bb1-4597-a20b-a6d72385523a

**Reply by danny-avila** [reply]
if the purpose is to show individual messages, why not just make use the Message component and structure the tree as desired based off that?

**Reply by danny-avila** [reply]
the ChatView component is made up of modular parts. in fact you can benefit from making your own "View" component from the same parts.

**Reply by ryohei-kamiya** [reply]
@danny-avila 
I think it's a practical choice to focus only on display at first, as enabling posting messages on the MindMap screen would have a significant impact on the existing code.
Considering we are focusing only on display for now, it can indeed be made simpler as you suggested.  I was thinking of using MessagesView as a reference because I want to include a scroll function for long messages. Is my understanding correct?

**Reply by ryohei-kamiya** [reply]
Maybe something like this?

https://github.com/ryohei-kamiya/LibreChat/compare/70f2ad3...9ce8ff2

↓test

https://github.com/danny-avila/LibreChat/assets/2719533/c9baaa33-fc30-4fc8-8824-3022907b4556

**Reply by danny-avila** [reply]
nice yeah, I think building off this approach would be cool, especially with some way to reply to a specific thread

**Reply by ryohei-kamiya** [reply]
I've looked PR #1696 and understand the plan for the "SidePanel".
I can place the button to switch to MindMap view in the "SidePanel" as shown in the following image.
For now, I will proceed with the development of MindMapView assuming this specification.

<img width="701" alt="screenshot_20240204" src="https://github.com/danny-avila/LibreChat/assets/2719533/b76c60d8-b8ac-477a-8cd7-ae0f68492a0e">

**Reply by ryohei-kamiya** [reply]
@danny-avila 
Hi!
I've added "Views" to the SidePanel, allowing switching between the Chat view and the MindMap view.

↓Light mode
<img width="369" alt="sidepanel-views_light" src="https://github.com/danny-avila/LibreChat/assets/2719533/383d28f4-b540-4d01-be26-ee8d5101f114">

↓Dark mode
<img width="369" alt="sidepanel-views_dark" src="https://github.com/danny-avila/LibreChat/assets/2719533/651820db-a612-4872-9f63-46ca2ebeef45">

Additionally, in the MindMap view, I've implemented a restriction that prevents not only sending messages but also editing them. I plan to lift this restriction once the ability to send multiple messages simultaneously is implemented, making it possible to send and edit messages in the MindMap view as well. The restriction that prevents editing messages in the MindMap view was implemented by copying `components/Chat/Messages` to `components/MindMap/Messages` and hiding the SiblingSwitch and HoverButtons through commenting them out.

Furthermore, I've added the `reactflow` and `@dagrejs/dagre` dependency packages to `client/package.json`.

↓ Test

https://github.com/danny-avila/LibreChat/assets/2719533/f2b81a1e-c49a-4268-887b-991a0350ca20


↓ Difference
https://github.com/ryohei-kamiya/LibreChat/compare/637a1a4...2260be6

## Question about next action
Is this implementation OK?

If you accept this proposal, I'll clean up the commits and create a PR.
Also, I exclude the commit of `package-lock.json` updating in the PR.

**Reply by thonic** [reply]
I would find this feature interesting, too, are there any news?
